### ISSUE TITLE
Fuzz IOS applications.

### ISSUE THREAD

**Issue by Picasso-r** [AFLplusplus/AFLplusplus#1869]
Now I'm trying to use AFL++ to fuzz app on real IOS devices. I have tested the [fpicker](https://github.com/ttdennis/fpicker), but it only succeed on simulator in Xcode.
I refer to the issue and find that the afl_proxy is mentioned. Troubling with this for days, I'm here asking for some help.
**Problem** :
[A successful example on android goes](https://github.com/marcinguy/fpicker-aflpp-android) like that I should put AFL++ on IOS device. How can I do this? I mean there's nothing like adb, so how can I compile a binary on macos and 'push' it to IOS device and excute it on IOS device? Is there a terminal I can easily use or I should utilize ssh to connect iphone? Or more easily, suppose(actually I do) now I compile a binary with command lines on my mac which can be run on IOS device, how can I 'adb' it to IOS device and use it just like '$ ~ ./test ' .

**Comment by domenukk** [comment]
These are not really questions that we can answer as AFL++ devs I'm afraid. Feel free to add some documentation as PR if you find it out.

**Comment by musnows** [comment]
Hi, I'm trying to fuzz ios object-c code, could I do it with fpicker in ios simulator?  @Picasso-r

**Comment by Picasso-r** [comment]
Yes it works.But a faster way is to use libfuzzer. @musnows

**Comment by musnows** [comment]
> Yes it works.But a faster way is to use libfuzzer. @musnows 

I try libfuzzer before, getting error of libfuzzer .so not found on arm mac device, I don t know how to fix it.

Have u counter this issue before? https://github.com/ttdennis/fpicker/issues/43

**Comment by Picasso-r** [comment]
@musnows I think I met this problem before, but I forget the details. You can debug it step by step and see the log of fpicker. For libfuzzer, you can download its source code in llvm project and compile on your own, and that's why it's faster.

**Comment by musnows** [comment]
> [@musnows](https://github.com/musnows) I think I met this problem before, but I forget the details. You can debug it step by step and see the log of fpicker. For libfuzzer, you can download its source code in llvm project and compile on your own, and that's why it's faster.

Thx for your information, I will try it later!

**Comment by musnows** [comment]
@Picasso-r   Hi,  I could compile llvm_project and get libclang_rt.fuzzer_iossim.a with following command on macOS 15 arm

```sh
git clone https://github.com/llvm/llvm-project.git
cd llvm-project
mkdir -p build && cd build

cmake -G Ninja \
  -DCMAKE_BUILD_TYPE=Release \
  -DLLVM_ENABLE_PROJECTS="clang;compiler-rt" \
  -DCLANG_ENABLE_PROTO_FUZZER=OFF \
  -DLLVM_USE_SYSTEM_PROTOBUF=OFF \
  -DLLVM_ENABLE_PROTOBUF=OFF \
  -DLLVM_ENABLE_LIBFUZZER=ON \
  -DCOMPILER_RT_BUILD_LIBFUZZER=ON \
  -DLLVM_TARGETS_TO_BUILD="ARM;AArch64" \
  -DCMAKE_OSX_ARCHITECTURES="arm64" \
  -DLLVM_DEFAULT_TARGET_TRIPLE="aarch64-apple-ios" \
  ../llvm

ninja
```

getting so like this
```
➜  llvm-project find build -name "libclang_rt.fuzzer*"
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_no_main_iossim.a
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_interceptors_iossim.a
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_no_main_ios.a
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_ios.a
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_interceptors_osx.a
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_iossim.a
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_interceptors_ios.a
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_osx.a
build/lib/clang/22/lib/darwin/libclang_rt.fuzzer_no_main_osx.a
```

But I dont know how to intergate libclang_rt.fuzzer_iossim.a into iOS xcode project. AI told me that I need to create a target for macOS command line tools in the xcode project, and then call my object-c tested code. Is that how it works? Can libfuzz be directly integrated into an iOS test target?

**Comment by Picasso-r** [comment]
I just drag all libfuzzer source code files into xcode and compile it together with fuzz target into a whole app. I didn't succeed linking it as a library either.

**Comment by musnows** [comment]
> I just drag all libfuzzer source code files into xcode and compile it together with fuzz target into a whole app. I didn't succeed linking it as a library either.

Thx, so I need to copy all codes in `llvm-project/compiler-rt/lib/fuzzer`, add then as a source code for my iOS demo app, and compile it together? Do I need to directly add the source code to the target of the app demo or to the target of app unit test?

**Comment by vanhauser-thc** [comment]
@musnows stop hijacking this closed issue, especially on non-afl++ topics, otherwise I will need to ban you, sorry

**Comment by musnows** [comment]
> [@musnows](https://github.com/musnows) stop hijacking this closed issue, especially on non-afl++ topics, otherwise I will need to ban you, sorry

Sorry for that, wont do it again.

**Comment by vanhauser-thc** [comment]
@musnows seek help on the discord instead: https://github.com/AFLplusplus/AFLplusplus/issues/681

**Comment by Noi98** [comment]
> Now I'm trying to use AFL++ to fuzz app on real IOS devices. I have tested the [fpicker](https://github.com/ttdennis/fpicker), but it only succeed on simulator in Xcode. I refer to the issue and find that the afl_proxy is mentioned. Troubling with this for days, I'm here asking for some help. **Problem** : [A successful example on android goes](https://github.com/marcinguy/fpicker-aflpp-android) like that I should put AFL++ on IOS device. How can I do this? I mean there's nothing like adb, so how can I compile a binary on macos and 'push' it to IOS device and excute it on IOS device? Is there a terminal I can easily use or I should utilize ssh to connect iphone? Or more easily, suppose(actually I do) now I compile a binary with command lines on my mac which can be run on IOS device, how can I 'adb' it to IOS device and use it just like '$ ~ ./test ' .

**Comment by Noi98** [comment]
I tried to do fuzzing using Fpicker for iOS applications, but I encountered several issues. I want you to help me with them.![image](https://github.com/user-attachments/assets/5dfd563c-395b-4cab-9f57-456e4debffbf)

And this is the custom harness I created. Could there be mistakes in it that caused this error? I tried changing the FRIDA version, but it didn’t help.

![image](https://github.com/user-attachments/assets/21e41c93-f40f-43dc-86cf-60771fc16119)
@Picasso-r @musnows
